# Karpov courses

### Lesson_2

источник доп информации - документация pandas: pandas.DataFrame/ атрибуты и методы 

In [143]:
import pandas as pd
from datetime import datetime

In [15]:
df = pd.read_csv('d:COURSES/Анатолий Карпов/Твой путь в мир Data Analytics. Часть 1 (2020)/2/Лекции/lesson_1_data.csv',
                 encoding='windows-1251',
                 sep = ';')

In [16]:
df

Номер     Дата создания       Дата оплаты  \
0    1062823  01.12.2019 10:50  01.12.2019 10:52   
1    1062855  01.12.2019 20:53  01.12.2019 21:27   
2    1062856  01.12.2019 21:43               NaN   
3    1062880   03.12.2019 0:18               NaN   
4    1062899  03.12.2019 21:43               NaN   
..       ...               ...               ...   
287  1064720   30.12.2019 9:42  30.12.2019 12:49   
288  1064724  30.12.2019 11:32               NaN   
289  1064775   31.12.2019 2:17   31.12.2019 2:22   
290  1064793  31.12.2019 16:40  01.01.2020 14:29   
291  1064796  31.12.2019 17:29  31.12.2019 17:32   

                                                 Title    Статус  Заработано  \
0                              Курс обучения «Эксперт»  Завершен    29597.50   
1                              Курс обучения «Эксперт»  Завершен    17450.30   
2                           Курс обучения «Специалист»   Отменен        0.00   
3                          Курс обучения «Консультант»   Отменен        0.00   
4                              Курс обучения «Эксперт»   Отменен        0.00   
..                                                 ...       ...         ...   
287  Подписка «ОНЛАЙН ДИЕТОЛОГ» с ежемесячным автос...  Завершен     2935.44   
288                        Курс обучения «Консультант»   Отменен        0.00   
289                        Курс обучения «Консультант»  Завершен     7423.92   
290  Подписка «ОНЛАЙН ДИЕТОЛОГ» с ежемесячным автос...  Завершен     2935.44   
291  Курс от Школы Диетологов. Повышение квалификац...  Завершен     9898.56   

                             Город                 Платежная система  
0                             Чита                Сбербанк эквайринг  
1                        Краснодар                      Яндекс.Касса  
2                              NaN                               NaN  
3    г.Москва и Московская область                               NaN  
4    г.Москва и Московская область                               NaN  
..                             ...                               ...  
287              Самарская область                      Яндекс.Касса  
288                            NaN                               NaN  
289                            NaN  Сбербанк эквайринг,Бонусный счет  
290             Республика Карелия                      Яндекс.Касса  
291                            Уфа  Сбербанк эквайринг,Бонусный счет  

[292 rows x 8 columns]

In [27]:
df.head() # пять или иное () кол-во значений сверху таблицы

Номер     Дата создания       Дата оплаты                        Title  \
0  1062823  01.12.2019 10:50  01.12.2019 10:52      Курс обучения «Эксперт»   
1  1062855  01.12.2019 20:53  01.12.2019 21:27      Курс обучения «Эксперт»   
2  1062856  01.12.2019 21:43               NaN   Курс обучения «Специалист»   
3  1062880   03.12.2019 0:18               NaN  Курс обучения «Консультант»   
4  1062899  03.12.2019 21:43               NaN      Курс обучения «Эксперт»   

     Статус  Заработано                          Город   Платежная система  
0  Завершен     29597.5                           Чита  Сбербанк эквайринг  
1  Завершен     17450.3                      Краснодар        Яндекс.Касса  
2   Отменен         0.0                            NaN                 NaN  
3   Отменен         0.0  г.Москва и Московская область                 NaN  
4   Отменен         0.0  г.Москва и Московская область                 NaN

 - пять или иное () кол-во значений СВЕРХУ таблицы

In [26]:
df.tail() # пять или иное () кол-во значений снизу таблицы

Номер     Дата создания       Дата оплаты  \
287  1064720   30.12.2019 9:42  30.12.2019 12:49   
288  1064724  30.12.2019 11:32               NaN   
289  1064775   31.12.2019 2:17   31.12.2019 2:22   
290  1064793  31.12.2019 16:40  01.01.2020 14:29   
291  1064796  31.12.2019 17:29  31.12.2019 17:32   

                                                 Title    Статус  Заработано  \
287  Подписка «ОНЛАЙН ДИЕТОЛОГ» с ежемесячным автос...  Завершен     2935.44   
288                        Курс обучения «Консультант»   Отменен        0.00   
289                        Курс обучения «Консультант»  Завершен     7423.92   
290  Подписка «ОНЛАЙН ДИЕТОЛОГ» с ежемесячным автос...  Завершен     2935.44   
291  Курс от Школы Диетологов. Повышение квалификац...  Завершен     9898.56   

                  Город                 Платежная система  
287   Самарская область                      Яндекс.Касса  
288                 NaN                               NaN  
289                 NaN  Сбербанк эквайринг,Бонусный счет  
290  Республика Карелия                      Яндекс.Касса  
291                 Уфа  Сбербанк эквайринг,Бонусный счет

- пять или иное () кол-во значений СНИЗУ таблицы

т.к. номер последнего индекса 291 => всего наблюдений в таблице 292 (т.к. первый индекс нумеруется с 0)

In [31]:
df.shape

(292, 8)

- атрибут, выдает размерность нашей таблицы: (строки, столбцы)

In [33]:
df.dtypes

Номер                  int64
Дата создания         object
Дата оплаты           object
Title                 object
Статус                object
Заработано           float64
Город                 object
Платежная система     object
dtype: object

- атрибут, выдающий типы данных в колонках

In [36]:
df.describe()

Номер    Заработано
count  2.920000e+02    292.000000
mean   1.063745e+06   3397.615034
std    4.438688e+02   5771.572829
min    1.062823e+06      0.000000
25%    1.063608e+06      0.000000
50%    1.063698e+06   2935.440000
75%    1.063807e+06   2935.440000
max    1.064796e+06  42750.000000

- метод по описанию (расчету) основных показателей для числовых значений таблицы

### Приступаем к Предподкотовке данных в подгруженном датафрейме. 
1. Переименовываем названия колонок в питоновский вид, чтобы все на англ и без пробелов (только через "_")

In [56]:
df.columns

Index(['Nmb', 'Date_create', 'Date_coast', 'title', 'Status', 'Profit', 'City',
       'Pay_sistem'],
      dtype='object')

In [55]:
df = df.rename(columns={'Номер': 'number',
                        'Дата создания': 'create_date',
                        'Дата оплаты': 'payment_date',
                        'Title' : 'title',
                        'Статус':'status', 
                        'Заработано': 'profit',
                        'Город': 'city',
                        'Платежная система': 'pay_sistem'})

повторяю переименование ниже, т.к. поторопился с переименованием, нужно подправить:

In [66]:
df = df.rename(columns={'Nmb': 'number',
                        'Date_create': 'create_date',
                        'Date_coast': 'payment_date',
                        'Title' : 'title',
                        'Status':'status', 
                        'profit': 'money',
                        'City': 'city',
                        'pay_sistem': 'payment_system'})

In [67]:
df.columns

Index(['number', 'create_date', 'payment_date', 'title', 'status', 'money',
       'city', 'payment_system'],
      dtype='object')

In [68]:
df.head()

number       create_date      payment_date                        title  \
0  1062823  01.12.2019 10:50  01.12.2019 10:52      Курс обучения «Эксперт»   
1  1062855  01.12.2019 20:53  01.12.2019 21:27      Курс обучения «Эксперт»   
2  1062856  01.12.2019 21:43               NaN   Курс обучения «Специалист»   
3  1062880   03.12.2019 0:18               NaN  Курс обучения «Консультант»   
4  1062899  03.12.2019 21:43               NaN      Курс обучения «Эксперт»   

     status    money                           city      payment_system  
0  Завершен  29597.5                           Чита  Сбербанк эквайринг  
1  Завершен  17450.3                      Краснодар        Яндекс.Касса  
2   Отменен      0.0                            NaN                 NaN  
3   Отменен      0.0  г.Москва и Московская область                 NaN  
4   Отменен      0.0  г.Москва и Московская область                 NaN

In [69]:
df['title']

0                                Курс обучения «Эксперт»
1                                Курс обучения «Эксперт»
2                             Курс обучения «Специалист»
3                            Курс обучения «Консультант»
4                                Курс обучения «Эксперт»
                             ...                        
287    Подписка «ОНЛАЙН ДИЕТОЛОГ» с ежемесячным автос...
288                          Курс обучения «Консультант»
289                          Курс обучения «Консультант»
290    Подписка «ОНЛАЙН ДИЕТОЛОГ» с ежемесячным автос...
291    Курс от Школы Диетологов. Повышение квалификац...
Name: title, Length: 292, dtype: object

In [75]:
df[ ['title', 'status'] ].head()

title    status
0      Курс обучения «Эксперт»  Завершен
1      Курс обучения «Эксперт»  Завершен
2   Курс обучения «Специалист»   Отменен
3  Курс обучения «Консультант»   Отменен
4      Курс обучения «Эксперт»   Отменен

- пример вызова двух колонок (нужно было обрамить их в список [] ) только первых 5-ти сверху

In [79]:
df.status

0      Завершен
1      Завершен
2       Отменен
3       Отменен
4       Отменен
         ...   
287    Завершен
288     Отменен
289    Завершен
290    Завершен
291    Завершен
Name: status, Length: 292, dtype: object

- вариант вызова колонки - работает если название колонки без пробелов

По завершению работы с данными, перед возвращением данных заказчику при необходимости можно вернуть исконные названия - все в наших руках и ради удобства читабельности.

### Приступаем к анализу

Вводная 1: 

Отдел продаж: помогите нам составить отчет по тому - какие курсы нам принесли сколько денег, отсортировать по зароботку и разбить по городам.  

Лайфхак_1: Создаем переменную с общей стоимостью, для итогового сравнения наших расчетов с итоговой суммой. (самопроверка)

In [84]:
all_money = df.money.sum()
print(all_money)

992103.5900000001


Теперь переходим к анализу

1.1. Формулирование задачи на русском:

    Мы хотим найти суммарное количество всех денег, по каждому из курсов. Проранжировать по курсы по степени кол-ва заработка денег + посмотреть разбивку по городам


In [109]:
money_by_city = df.groupby(['title', 'city'], as_index=False) \
    .aggregate({'money': 'sum'}) \
    .sort_values('money', ascending=False)

money_by_city

title  \
51                             Курс обучения «Эксперт»   
156  Подписка «ОНЛАЙН ДИЕТОЛОГ» с ежемесячным автос...   
39                             Курс обучения «Эксперт»   
95   Подписка «ОНЛАЙН ДИЕТОЛОГ» с ежемесячным автос...   
30                          Курс обучения «Специалист»   
..                                                 ...   
63   Курс от Школы Диетологов. Повышение квалификац...   
61   Курс от Школы Диетологов. Повышение квалификац...   
55                    Курс от Школы Диетологов. Бизнес   
54                    Курс от Школы Диетологов. Бизнес   
158  Подписка «ОНЛАЙН ДИЕТОЛОГ» с ежемесячным автос...   

                                          city     money  
51   г.Санкт-Петербург и Ленинградская область  59195.00  
156              г.Москва и Московская область  46967.04  
39                                      Балхаш  42750.00  
95                          Краснодарский край  38169.78  
30                                   Краснодар  29695.70  
..                                         ...       ...  
63                                    Кемерово      0.00  
61                               Екатеринбург       0.00  
55                              Крым Советский      0.00  
54                                        Киев      0.00  
158                            нижний Новгород      0.00  

[159 rows x 3 columns]

- groupby('*', as_index=False) - присваивает и выводит доп индекс перечисляемым строкам

- \   - символ разделения строки для читабельности

- aggregate - вызов агрегации (какого-то метода обработки диапазона)

- sort_values ('*', ascending = False) - сортировка по убыванию

Выгрузка результата работы, в csv:


In [110]:
money_by_city.to_csv('money_by_city.csv', index= False)

Вводная 1.2: 

Посчитать сколько суммарно денег и завершено заказов. Какое кол-во заказов позволило заработать кол-во денег. Без групировки по городам

In [140]:
money_by_title = df \
    .query(" status == 'Завершен' ") \
    .groupby('title', as_index=False) \
    .aggregate({'money': 'sum', 'number': 'count'}) \
    .sort_values('money', ascending=False) \
    .rename(columns = {'number': 'success_orders'})

money_by_title

title      money  \
5  Подписка «ОНЛАЙН ДИЕТОЛОГ» с ежемесячным автос...  366947.20   
0                        Курс обучения «Консультант»  208163.49   
1                         Курс обучения «Специалист»  160862.64   
2                            Курс обучения «Эксперт»  148992.80   
4  Курс от Школы Диетологов. Повышение квалификац...   88384.92   
3                   Курс от Школы Диетологов. Бизнес   18752.54   

   success_orders  
5             125  
0              31  
1               7  
2               5  
4               9  
3               3

.query(" status == 'Завершен' ") - доп фильтр (условие отбора) 

.rename(columns = {'number': 'success_orders'}) - переименование колонки (выполнить в конце после вычислений)

Проверяем не потеряли ли мы данные, к примеру по общей стоимости:

- сравниваем результат с заранее заготовленной переменной all_money

In [141]:
if int(money_by_title.money.sum()) == int(all_money):
    print('Молодчага, без проебов!')
else:    
    print('Пиздэйс, ошибочкос, проверяй - исправляй, бро!')

Молодчага, без проебов!


Проверяем количество курсов, не потеряли ли что:
- сравниваем кол-во уникальных строк в title и в итоговом        

In [129]:
df.title.unique()

array(['Курс обучения «Эксперт»', 'Курс обучения «Специалист»',
       'Курс обучения «Консультант»', 'Курс от Школы Диетологов. Бизнес',
       'Курс от Школы Диетологов. Повышение квалификации. ',
       'Подписка «ОНЛАЙН ДИЕТОЛОГ» с ежемесячным автосписанием'],
      dtype=object)

In [130]:
money_by_title

title      money  \
5  Подписка «ОНЛАЙН ДИЕТОЛОГ» с ежемесячным автос...  366947.20   
0                        Курс обучения «Консультант»  208163.49   
1                         Курс обучения «Специалист»  160862.64   
2                            Курс обучения «Эксперт»  148992.80   
4  Курс от Школы Диетологов. Повышение квалификац...   88384.92   
3                   Курс от Школы Диетологов. Бизнес   18752.54   

   success_orders  
5             125  
0              31  
1               7  
2               5  
4               9  
3               3

  Добавляем к скрипту выдачу текущей даты и присвоения имени файла текущей даты (для поиска решения загуглили вопрос):

from datetime import datetime     --- строку с имортом для корректности отправлили в самый верх файла

datetime.today().strftime('%Y-%m-%d')

In [158]:
money_by_title = df \
    .query(" status == 'Завершен' ") \
    .groupby('title', as_index=False) \
    .aggregate({'money': 'sum', 'number': 'count'}) \
    .sort_values('money', ascending=False) \
    .rename(columns = {'number': 'success_orders'})

In [159]:
today_day = datetime.today().strftime('%Y-%m-%d')
file_name = 'money_by_title_{}.csv'
file_name = file_name.format(today_day)

In [160]:
if int(money_by_title.money.sum()) == int(all_money):
    print('ОК! File {} is written.'.format(file_name))
    money_by_title.to_csv(file_name, index=False)
else:    
    print('Пиздэйс, ошибочкос, проверяй - исправляй, бро!')

ОК! File money_by_title_2024-07-16.csv is written.
